upf/quantifiers_remover.py

```diff
@@ -0,0 +1,174 @@
+        #product([1,2], [3,4]) = (1,3) (1,4) (2,3) (2,4)
+        possible_combinations = list(product(*possible_objects))
+        subs_results = []
+        for o in possible_combinations:
+            ol: List[Object] = list(o)
+            subs: Dict[Expression, Expression] = dict(zip(vars, ol))
+            subs_results.append(self._substituter.substitute(args[0], subs))
+        return self._env.expression_manager.And(subs_results)
+
+
+class QuantifiersRemover():
+    '''Conditional effect remover class:
+    this class requires a problem and offers the capability
+    to transform a problem with quantifiers into a problem without.
+    '''
+    def __init__(self, problem: Problem):
+        self._problem = problem
+        self._action_mapping: Dict[Action, Action] = {}
+        self._env = problem.env
+        self._counter: int = 0
+        self._noquantifier_problem = None
+        #NOTE no simplification are made. But it's possible to add them in key points
+        self._simplifier = Simplifier(self._env)
+        self._expression_quantifier_remover = ExpressionQuantifierRemover(self._problem, self._env)
+
+    def get_rewritten_problem(self) -> Problem:
+        '''Creates a problem that is a copy of the original problem
+        but every quantifier is removed.'''
+        if self._noquantifier_problem is not None:
+            return self._noquantifier_problem
+        #NOTE that a different environment might be needed when multy-threading
+        new_problem = self._create_problem_copy_without_actions_and_goals()
+        for a in self._problem.actions().values():
+            na = self._action_without_quantifiers(a)
+            new_problem.add_action(na)
+            self._action_mapping[na] = a
+        for g in self._problem.goals():
+            ng = self._expression_quantifier_remover.remove_quantifiers(g)
+            new_problem.add_goal(ng)
+        return new_problem
+
+    def _create_problem_copy_without_actions_and_goals(self):
+        '''Creates the shallow copy of a problem, without adding the actions
+        with quantifiers and by pushing them to the stack
+        '''
+        new_problem: Problem = Problem("noquantifiers_" + str(self._problem.name()), self._env)
+        for f in self._problem.fluents().values():
+            new_problem.add_fluent(f)
+        for o in self._problem.all_objects():
+            new_problem.add_object(o)
+        for fl, v in self._problem.initial_values().items():
+            new_problem.set_initial_value(fl, v)
+        return new_problem
+
+    def _action_without_quantifiers(self, action) -> Action:
+        #emulates a do-while loop: searching for an available name
+        is_unavailable_name = True
+        while is_unavailable_name:
+            new_action_name = action.name()+ "_" +str(self._counter)
+            self._counter = self._counter + 1
+            is_unavailable_name = self._problem.has_action(new_action_name)
```

Comment:
Why do you need to rename the actions? The new actions are add to the `new_problem` so keeping their original name should not be a problem, right?

Reply:
Yes this is totally true. This would also remove the need of a "write back" structure.

This was a remaining of the copy-paste from the ConditionalEffectsRemover